genienlp/run_generation.py

```diff
@@ -201,76 +231,208 @@ def sample_sequence(model, length, context, position_ids, num_samples=1, tempera
+            all_input_sequences.append(input_sequence)
+            all_input_sequence_lengths.append(len(input_sequence_tokens))
+            all_context_tokens.append(context_tokens)
+            all_context_lengths.append(len(context_tokens))
+
+    return all_input_sequences, all_input_sequence_lengths, all_context_tokens, all_context_lengths, all_golds, reverse_maps
+
+
+def is_question(sentence: str):
+    question_words = ['which', 'what', 'where', 'how', 'who', 'when', 'is', 'are', 'am', \
+                      'can', 'could', 'would', 'will', 'have', 'did', 'do', 'does', 'no is', 'yes is']
+    for w in question_words:
+        if sentence.startswith(w+' '):
+            return True
+    return False
+
+def input_heuristics(s: str, thingtalk=None, is_cased=False, keep_special_tokens=False, keep_tokenized=False):
+    """
+    Changes the input string so that it is closer to what the pre-trained language models have seen during their training.
     Outputs:
         s: the new string
         reverse_map: a list of special tokens. Can be used to recover the original special_tokens in the string
     """
     reverse_map = []
     s = s.strip()
-    s = detokenize(s)
+    s = tokenize(s)
 
     # Put question mark at the end whenever necessary.
-    if s.startswith('which') or s.startswith('what') or s.startswith('where') or s.startswith('how') or s.startswith('who') or s.startswith('when'):
-        if s.endswith('.'):
-            s = s[:-1]
-        s += '?'
+    sentences = [sentence.strip() for sentence in re.split('\s+([.?!:])\s*', s) if len(sentence) > 0]
+    # print('sentences = ', sentences)
+    for idx in range(len(sentences)):
+        if sentences[idx] in ['.', '?' , '!', ':']:
+            continue
+        if idx == len(sentences)-1 or sentences[idx+1] not in ['.', '?', '!', ':']:
+            # add the missing punctuation
+            if is_question(sentences[idx]):
+                sentences[idx] = sentences[idx] + '?'
+            else:
+                sentences[idx] = sentences[idx] + '.'
+        else:
+            if is_question(sentences[idx]):
+                assert sentences[idx+1] in ['.', '?', '!', ':']
+                sentences[idx+1] = '?'
+
+        if is_cased:
+            # capitalize the first word and parameters
+            if thingtalk:
+                _, parameters = remove_thingtalk_quotes(thingtalk)
+                # print('parameters = ', parameters)
+                for p in parameters:
+                    capitalized_p = ' '.join([t[0].upper()+t[1:] for t in p.split()])
+                    sentences[idx] = sentences[idx].replace(p, capitalized_p)
+            sentences[idx] = sentences[idx].replace(' i ', ' I ')
+            sentences[idx] = sentences[idx][0].upper()+sentences[idx][1:]
+            
+    s = ' '.join(sentences)
+    if not keep_tokenized:
+        s = detokenize(s)
+    
+    if not is_cased:
+        s = lower_case(s)
```

Comment:
This file is _definitely_ not the place for this kind of (language-specific) heuristics. Either move them to a preprocessing of the dataset, or somewhere else.